Updates for Tcl/Tk 9.0 support, depends upon PR #22

mark987 opened the pull request:
I've been working on tcl 8.6 / 9.0 updates for a while, have read the proposed patches from @qykth-git and I believe these miss some important things, it is easier for me to just show my work including explanations in the commit messages. Key differences:

There are a number of updates to mouse scroll handling we should make for Tcl/Tk 8.6, these depend upon TIP 171, and this precedes anything for TIP 474 / Tcl 9. We should update gitk requirements to be 8.6 before doing this as 8.4/8.5 don't support TIP 171, 8.6 will likely continue in use for a decade as 9.0 just hit. My 9.0 mouse version uses the Option- modifier to select large motion (like on x11 previously) vs fine motion (line on aqua previously), similar to how Tk changes this by default on other widgets.

Tcl 9 includes TIP 699 because of much confusion about binary channels. Following that document and looking carefully at gikt, I reach a different result than 042ede8, reasoning is in the 0f4b70cc2d commit message.

There are additional changes using -profile needed to operate on Tcl 9 to cover some encoding issues, 5d5d337bda and 036fc90675.

I'm successfully using gitk (and git-gui) on Tcl8.6 and on Tcl9.0.1 with this set (no longer have 8.7, but did some of the earlier work with that).

mark987 commented:
I've updated the branch to:
- correct spelling errors and generally improve wording in commit messages
- moved settting and use of `colspc` variable into 432717b from its successor (now 61da1f5)
- increased the value of colspc from 5 to 25, horizontal scrolling of the commit graph is now 5x larger than before.

To @j6t on your previous questions:
horizontal scrolling of the commit graph is a regression on MacOS
- I updated comments in 432717b about this: basically, normal GUI design has the mouse pointer indicate the window to be scrolled. While vertical scolling in any of the top 3 panes makes sense, as all scroll, horizontal scrolling of the commit graph from any pane does not as ONLY the commit graph will scroll.

I am not happy with the simple tilde substitution with $env(HOME)
- What I wrote works on all the platforms I know how to test and can explain. I know the configuration of git on Windows using either Cygwin nor MSYS (the latter underlies g4w). MSYS defines HOME, g4w assures this is propagated, but not of other approaches.
- Tilde expansion on Tcl 9 uses $HOME if it is defined. Tcl sets up the environment when starting, and tries to assure $HOME is defined, using $HOMEDRIVE and $HOMEDIR if $HOME is not there. But, those variables are wrong on Cygwin and g4w, so I'm not at all sure what they mean.
- So, long as $HOME is honored and not touched, this doesn't matter to me. I just don't understand a supportable configuration where $HOME is not defined.

#11 changes fconfigure to chan configure. Is this not needed for Tcl 9?
- The "chan" command was introduced in Tcl 8.5 per TIP 208 to consolidate many (by no means all) channel related commands into one group. A few commands, including fconfigure, were recommended to be deprecated in TIP 208, but this did not happen in 8.5 or 8.6. In 9.0, fconfigure and chan configure are the same thing, though the Tcl 9 documentation for fconfigure shows this is now deprecated. Nothing in the documentation on migrating to Tcl 9 mentions this, and fconfigure shows up in the migration instructions.
- My take is this is optional, not required. We could do a global search/replace, it makes no difference. So, I didn't do it.

horizontal scrolling on X11 now works secretly with the mouse wheel (Shift+Alt+wheel), but it uses a very small increment here.
- The colspc variable mentioned above can be increased to make this scroll more quickly, I changed it from 5 to 25, giving a 5x increase.
- We might also use $linespc instead of $colspc to set horizontal scolling increments : linespc is computed as the vertical font size, so is roughly the size of a line, this could be a reasonable value to use for horizontal scrolling as well.
- You have to use Tcl9 to see any difference when the Alt key is pressed: Tcl8 does not generate the Option- modifier.

qykth-git commented:
@j6t 
`$env(HOME)` was defined like this on Windows Tcl.
https://github.com/tcltk/tcl/blob/9f8cc0d427b6fea52116a71060c9a37a757efea4/win/tclWinInit.c#L558

Windows Tcl build own `$env(HOME)` with these order.
1. Use supplied `$HOME`
2. Build `$HOME` from `$HOMEDRIVE` and `$HOMEPATH`
3. Build `$HOME` from `$USERPROFILE`
4. Just use `C:\\` as `$HOME`

j6t commented:
@mark987 Thank you for the update and the explanations. Using `$env(HOME)` for the tilde is alright. The horizontal scrolling speed is now a lot better.

The need for `-profile tcl8` is scary stuff. (Who the hell writes software that expects a perfect environment, and crashes at the first imperfection, I have to ask the Tcl people?) Do we need a plan to do something similar for the exec calls that capture output from the commands?

mark987 commented:
The solution actually removes lines of code, as we no longer need the Windows exec override :^)  Instead, just use a pipeline from [open | ..], and take advantage that we already overwrote open to add -profile tcl8.

j6t commented:
Thank you for the extensive tests.

The `exec` replacement that you suggest isn't something I would take. It's too scary. There could be library calls outside of Gitk that expect a certain API, but the suggested replacement isn't fully compatible.

At any rate, the important finding is that `exec` of neither Tcl 8.6 nor Tcl 9 fails on invalid UTF-8 input. That the two versions don't give the same result (or an "unexpected" result), is of no concern for Gitk. (After all, its a form of garbage-in-garbage-out.) Therefore, I wouldn't do anything about it at this time.

mark987 commented:
I've reworked the 8.6 scrolling, with the goal of making all platforms do the same thing (they did not come even close, it turns out, and for no good reason). I also wanted to make 8.6 and 9.0 do the same thing but failed. The problem is that 9.0 adds the Option- modifier using the Alt key on PC, and Option key on Mac, to get ~5x (or more) motion than the nominal. This key cannot be changed. But, in 8.6 and earlier, the Alt- key on PC/Windows activates the menu-bar, this cannot be changed, and makes the Alt- key as a modifier for mouse scrolling essentially unusable.

I added a mouse motion multiplier (kscroll) in the preferences dialog, default value = 1, allowed [1..20]. 1 gives essentially the old aqua motion, 5 the win32 motion, and the x11 had a mixture of those two. A value < 1 is not useful, though you can type in whatever fractional value you wish even though the spinbox only does integers.

So, in 8.6, on all platforms, the 5 main windows (3 top panes, two text widgets below) are scrolled vertically by kscroll lines (the text widgets are weird, move 2-3 lines when asked to move 1, can do nothing about that). Horizontal scrolling is available on the two left panes.

9.0 does the same, but adds the Option- modifier to get more motion. With this I set the kscroll preference to 1, use Alt- to move far. But, as I said, cannot get this for 8.6.

More lines of bindings than I wished, but this is the only way to get the uniform scrolling I had put into win32 15 years ago, and forgot all about if I even noticed it then :^).

mark987 commented:
I've incorporated your suggestion, along with a refactoring of the scrolling code. I was bothered about the number of mouse binding lines added, their replication, as well as incomplete details in the scrolling calculations. My new branch is mostly a refactoring to reduce duplication, but adds some features using a common [scroll_val ...] function
- the scroll val is integer, required on Tk 8.6, regardless of the user setting a floating point scolling multiplier
- the scroll value adjusts to accommodate the nominal minimum 3 line scrolling of text widgets.
- the default preference is now 3, compromise of the former behavior, and also the minimum value at which the canvas and text widgets scroll by the same amount.
Overall, fewer lines added, more understandable to me.

I don't have anymore nagging problems with this code, so won't change anything more except to address feedback. Tested on Linux/X11 with Tcl 8.6 and Tcl 9.0.1, Cygwin/X11 with Tcl 8.6.12, G4W/win32 with Tcl 8.6.13. I don't have Tcl 9 except on Linux, no access to aqua anywhere, so those are the big holes in testing.

j6t commented:
Two small things I noticed while going over my patch list today:

- The text "Wheel scrolling multiplier" is not prepared for translation. It should be.
- We have two variables `have_tk85` and `have_tk86`. They become obsolete with this series and all code using them can be simplified by assuming that the flags are true.

mark987 commented:
I've updated my branch with:                                                                
- scrubbed commit messages
- added [mc "Wheel scrolling multiplier"]
- changed mouse_wheel_scale to scroll_D0, as this is expected value of %D passed into the event
- moved the commit allowing Tcl9 until after all Tcl9 compatibility is added.
- added a number of additional commits removing compatibility with pre 8.6 Tcl/Tk and early git.

I was holding the latter patches back until the dust settles on the first, but, you asked :^) The first one, a3a7e5e, removes have_tk85 and have_tk86. I found much more to do, though. I've been running with all of these for a while.

j6t commented:
Thank you for the update.

Since I have a few topics in flight that change the user-interface, I would prefer to postpone the `ttk` changes until those topics are merged.

Regarding the requirement of a modern `git`, I of course agree that we remove compatibility for ancient versions. I would nevertheless target Debian [Buster](https://packages.debian.org/buster/git) as the minimum, which has Git 2.20. Looking at 8c7307e61012a, I don't think we must require 2.36.

mark987 commented:
The issue on git version traces to git-gui, gitk launches git-gui, so gitk logically needs git supporting gitk AND git-gui.
git added core.hooksPath in v2.10, and this broke git-gui until I patched git-gui for Junio's git 2.43 release. This relies upon git-hook-run added in v2.36. So, git-gui post 2.43 is compatible with git v2.36 and later, earlier git-gui is compatible only up to git 2.09. git 2.10 --> 2.35 is not compatible as hooks do not run (or the wrong hooks run, this got reported to me by a former colleague causing me to pursue fixing the problem).

So, git 2.20 in Debian Buster is in "no-man's" land vis-a-vis git-gui. For reasons above, I have trouble saying gitk is ok.

j6t commented:
Thanks, noted.
